R/utils-latency.R

```diff
@@ -0,0 +1,294 @@
+#' @param groups the grouping by which to fill forward
+#' @importFrom tidyselect all_of
+#' @importFrom rlang list2
+#' @importFrom vctrs vec_cast
+#' @importFrom dplyr across arrange bind_rows group_by summarise
+#' @keywords internal
+pad_to_end <- function(x, groups, end_date, columns_to_complete = NULL) {
+  if (is.null(columns_to_complete)) {
+    columns_to_complete <- setdiff(names(x), key_colnames(x))
+  }
+  itval <- epiprocess:::guess_period(c(x$time_value, end_date), "time_value")
+  # get the time values we need to fill in
+  completed_time_values <-
+    x %>%
+    group_by(across(all_of(groups))) %>%
+    summarise(
+      time_value = list2(
+        time_value = seq_null_swap(from = max(time_value) + itval, to = end_date, by = itval)
+      )
+    ) %>%
+    unnest("time_value") %>%
+    mutate(time_value = vec_cast(time_value, x$time_value))
+  # pull the last value in each group and fill forward
+  grouped_and_arranged <- x %>%
+    arrange(across(all_of(c("time_value", groups)))) %>%
+    group_by(across(all_of(groups)))
+  values_to_fill <- grouped_and_arranged %>%
+    slice(min(across(all_of(columns_to_complete), count_single_column)):n())
+  filled_values <- values_to_fill %>%
+    bind_rows(completed_time_values) %>%
+    arrange(across(all_of(c("time_value", groups)))) %>%
+    fill(all_of(columns_to_complete), .direction = "down") %>%
+    slice(-1) # remove the oirginal rows
+
+  grouped_and_arranged %>%
+    slice(1:min(across(all_of(columns_to_complete), count_single_column))) %>%
+    bind_rows(filled_values) %>%
+    arrange(across(all_of(key_colnames(x)))) %>%
+    ungroup() %>%
+    group_by(across(all_of(get_grouping_columns(x))))
+}
+
+#' return the names of the grouped columns, or `NULL`
+#' @param x an epi_df
+#' @keywords internal
+#' @importFrom utils head
+get_grouping_columns <- function(x) {
+  group_names <- names(attributes(x)$groups)
+  head(group_names, -1)
+}
+
+#' get the location of the last real value
+#' @param col the relevant column
+#' @param from_last instead of the number of columns
+#' @keywords internal
+count_single_column <- function(col, from_last) {
+  run_lengths <- rle(is.na(col))
+  n_el <- length(col)
+  if (tail(run_lengths$values, 1)) {
+    n_at_end <- tail(run_lengths$lengths, 1)
+    return(n_el - n_at_end)
+  } else {
+    return(n_el)
+  }
+}
+
+
+#' seq, but returns null if from is larger
+#' @keywords internal
+seq_null_swap <- function(from, to, by) {
+  if (from > to) {
+    return(NULL)
+  }
+  seq(from = from, to = to, by = by)
+}
+
+
+#' warn when the latency is larger than would be reasonable
+#' @param dataset the epi_df
+#' @param latency_table the whole collection of latencies
+#' @param target_columns the names of the columns that we're adjusting, and whether its unreasonably latent
+#' @keywords internal
+check_interminable_latency <- function(dataset, latency_table, target_columns, forecast_date, call = caller_env()) {
```

Comment:
If we're going to be producing these sorts of warnings, we should have a flag that turns it off.

Reply:
should also be in the canned forecaster